bindings/trycp_runner/src/common.rs

```diff
@@ -45,6 +56,253 @@ pub fn connect_trycp_client<SV: UserValuesConstraint>(
+pub fn install_app<SV>(
+    ctx: &mut AgentContext<TryCPRunnerContext, TryCPAgentContext<SV>>,
+    app_path: PathBuf,
+    role_name: &RoleName,
+) -> WindTunnelResult<()>
+where
+    SV: UserValuesConstraint,
+{
+    let run_id = ctx.runner_context().get_run_id().to_string();
+    let client = ctx.get().trycp_client();
+    let agent_id = ctx.agent_id().to_string();
+
+    let (app_port, cell_id, credentials) = ctx
+        .runner_context()
+        .executor()
+        .execute_in_place(async move {
+            let agent_key = client
+                .generate_agent_pub_key(agent_id.clone(), None)
+                .await?;
+
+            let content = std::fs::read(app_path)?;
+
+            let installed_app_id = format!("{}-app", agent_id).to_string();
+            let app_info = client
+                .install_app(
+                    agent_id.clone(),
+                    InstallAppPayload {
+                        source: AppBundleSource::Bundle(AppBundle::decode(&content)?),
+                        agent_key,
+                        installed_app_id: Some(installed_app_id.clone()),
+                        membrane_proofs: Default::default(),
+                        network_seed: Some(run_id),
+                    },
+                    None,
+                )
+                .await?;
+
+            let enable_result = client
+                .enable_app(agent_id.clone(), app_info.installed_app_id.clone(), None)
+                .await?;
+            if !enable_result.errors.is_empty() {
+                return Err(anyhow::anyhow!(
+                    "Failed to enable app: {:?}",
+                    enable_result.errors
+                ));
+            }
+
+            let app_port = client
+                .attach_app_interface(agent_id.clone(), None, AllowedOrigins::Any, None, None)
+                .await?;
+
+            let issued = client
+                .issue_app_auth_token(
+                    agent_id.clone(),
+                    IssueAppAuthenticationTokenPayload {
+                        installed_app_id,
+                        expiry_seconds: 30,
+                        single_use: true,
+                    },
+                    None,
+                )
+                .await?;
+
+            client
+                .connect_app_interface(issued.token, app_port, None)
+                .await?;
+
+            let cell_id = match app_info
+                .cell_info
+                .get(role_name)
+                .ok_or(anyhow::anyhow!("Role not found"))?
+                .first()
+                .ok_or(anyhow::anyhow!("Cell not found"))?
+            {
+                CellInfo::Provisioned(pc) => pc.cell_id.clone(),
+                _ => anyhow::bail!("Cell not provisioned"),
+            };
+            log::debug!("Got cell id: {:}", cell_id);
+
+            let credentials = client
+                .authorize_signing_credentials(
+                    agent_id.clone(),
```

Comment:
Originally I was confused thinking agent_id here was the agent pubkey, but no, this is an indicator to the multi client which sub-client we're talking to? I don't know if a different variable name would be clearer.

Also, is it true that since we set up the init with unrestricted access, this `authorize_signing_credentials` isn't strictly necessary... except that the calls need to be signed by *something*?

Reply:
Correct, it's nothing to do with the agent pubkey - I've renamed it to `agent_name` which doesn't imply a unique identifier and it hopefully clearer? What that field is for has documentation on the getter function if you're working in the code but I agree it's confusing once it's being reviewed or has been copied into a variable.

Correct that we have unrestricted zome call access but it's the zome calls from the scenario to the conductor that we're setting up signing credentials for. It's effectively pointless when we have access to the admin websocket anyway because we already have complete control over the conductor... but that's what the API requires. Hopefully now that it's hidden in a common function, scenario authors won't have to worry too much about that.